From 8c9b2ee7660d024002947be8bcc30abd77a9df06 Mon Sep 17 00:00:00 2001
From: Benjamin Coddington <bcodding@redhat.com>
Date: Tue, 6 Jan 2015 16:46:11 -0500
Subject: [fs] nfsd4: fix xdr4 count of server in fs_location4

Message-id: <c4cd3ebea6044aa221684388461e795ce004d2f1.1420562070.git.bcodding@redhat.com>
Patchwork-id: 102496
O-Subject: [RHEL7.1 PATCH fs BZ1164055 2/2] nfsd4: fix xdr4 count of server in fs_location4
Bugzilla: 1164055
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

Fix a bug where nfsd4_encode_components_esc() incorrectly calculates the
length of server array in fs_location4--note that it is a count of the
number of array elements, not a length in bytes.

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1164055
Upstream id: bf7491f1be5e125eece2ec67e0f79d513caa6c7e
Signed-off-by: Benjamin Coddington <bcodding@redhat.com>
Fixes: 082d4bd72a45 (nfsd4: "backfill" using write_bytes_to_xdr_buf)
Cc: stable@vger.kernel.org
Signed-off-by: J. Bruce Fields <bfields@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4xdr.c b/fs/nfsd/nfs4xdr.c
index d53d792..b345fef 100644
--- a/fs/nfsd/nfs4xdr.c
+++ b/fs/nfsd/nfs4xdr.c
@@ -1769,7 +1769,7 @@ static __be32 nfsd4_encode_components_esc(struct xdr_stream *xdr, char sep,
 
   str = end;
  }
- pathlen = htonl(xdr->buf->len - pathlen_offset);
+ pathlen = htonl(count);
  write_bytes_to_xdr_buf(xdr->buf, pathlen_offset, &pathlen, 4);
  return 0;
 }
-- 
1.7.1